From linux-usb-owner@vger.kernel.org  Tue Oct  5 15:58:49 2010
From: Alan Cox <alan@linux.intel.com>
Subject: usb: langwell_udc: cancel pending requests when controller is suspended.
To: linux-usb@vger.kernel.org
Date: Fri, 01 Oct 2010 14:59:16 +0100
Message-ID: <20101001135915.24212.57542.stgit@localhost.localdomain>

From: Philippe Skowronski <philippe.skowronski@intel.com>

It is safer to cancel pending requests before free dTD and dQH when
controller enters suspend state.

Signed-off-by: Philippe Skowronski <philippe.skowronski@intel.com>
Signed-off-by: Hao Wu <hao.wu@intel.com>
[Switch to spin_lock_irq as suggested by Alan Stern]
Signed-off-by: Alan Cox <alan@linux.intel.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/usb/gadget/langwell_udc.c |    5 +++++
 1 file changed, 5 insertions(+)

--- a/drivers/usb/gadget/langwell_udc.c
+++ b/drivers/usb/gadget/langwell_udc.c
@@ -3391,6 +3391,11 @@ static int langwell_udc_suspend(struct p
 	/* save PCI state */
 	pci_save_state(pdev);
 
+	spin_lock_irq(&dev->lock);
+	/* stop all usb activities */
+	stop_activity(dev, dev->driver);
+	spin_unlock_irq(&dev->lock);
+
 	/* free dTD dma_pool and dQH */
 	if (dev->dtd_pool)
 		dma_pool_destroy(dev->dtd_pool);
